### PR TITLE
Use `DeferredResult` class instead of `'async'` option

### PR DESCRIPTION
**Pull request by emcastillo**
Get rid of this flag and delegate the asynchronous declaration to the models.

### CODE REVIEW

**Comment by emcastillo** [comment]
Fixed! PTAL!

File: pytorch_pfn_extras/handler/_handler.py [hunk 1]
```diff
@@ -15,6 +16,39 @@
...
+    def done(self) -> bool:
+        """Returns `True` when a deferred result is available."""
+        raise NotImplementedError("done must be implemented")
+
+    def wait(self) -> 'DeferredResult':
+        """Does a blocking wait until the result is available."""
+        raise NotImplementedError("wait must be implemented")
+
+    def get(self) -> Optional[Any]:
+        """Returns the result when its available."""
```
Comment (kmaehashi) [inline pytorch_pfn_extras/handler/_handler.py:33]: Could describe the expected behavior when not done?

File: pytorch_pfn_extras/handler/_handler.py [hunk 2]
```diff
@@ -427,18 +454,25 @@ def train_step(
...
-        if self._async:
+        if isinstance(outs, DeferredResult):
             # async returns inmediately
             # Check if there is a completed iteration
             # If we enqueue everything first, we will blow up with
             # memory due to the dataloaders being in the background
             # We need to call the tagged runtimes since async mode
             # require different treatment depending on the device.
+            if len(self._ppe_modules) != 1:
+                print(self._ppe_modules)
```
Comment (kmaehashi) [inline pytorch_pfn_extras/handler/_handler.py:465]: Remove? :smile:

**Pull request merged**